### DISCUSSION TITLE
3D printing font

### DISCUSSION THREAD

**Discussion by evils** [fontforge/fontforge#4968] [General]
i'm posting this on the General category because it seems to fit in all 3 others, and there's nothing here yet :)

i want to make a font for [Fused Filament Fabrication 3D printing](https://en.wikipedia.org/wiki/Fused_filament_fabrication#Applications) small text into the first layer (photos below)

### Problem statement

with a 0.4mm nozzle, below 3mm regular fonts become quite tricky,
at that size, i've found [D-DIN](https://fontlibrary.org/en/font/d-din) ok, and am yet to try [TGL 0-17](https://fontlibrary.org/en/font/tgl)
very quickly thinner parts of strokes get below the 0.4mm width and features like dots tend to fail

this is further complicated by having to give the slicer a 3D model which it then converts to paths
(having a font that describes a print path directly is probably not easily usable directly,
but having a font like that would probably be a good starting point if anyone ever wanted to do that)
slicers have ways of making features thinner than the printable line width, but these tend to not work well on text

[Relief-SingleLine](https://github.com/isdat-type/Relief-SingleLine) is interesting but its focus on subtractive rather than additive manufacturing makes it not significantly better than other fonts for this application

(messiness around the start and end of a line makes having line print direction information useful, if that can be used)

it's possible to only print the negative space (kind of a debossing, possibly to be filled with paint later),
having printable counters is even more important when doing that

currently i have to determine how many units the line width should be at the size i want to print the font
then expand the lines during drawing to make sure the spacing between features is right and to see what it looks like
(and undo that every time i make an adjustment)
ideally, i could keep the project as lines, but live preview what the expansion would look like
and maybe even have some tooling help in adjusting for different widths
(which is equivalent to keeping the same width and changing the printed size of the font)

### Desired Features
a font which:
* can be printed with a 0.4mm nozzle and smaller than regular fonts without slicer thin line detection
* maintains
  * a specified line width
  * spacing between lines (no 2 lines touch)
  * printable counters
  * room for a stroke between characters
* can be scaled while maintaining the above
* has a complementary font describing how to fill the counters / fill and convexify the characters' outline
* suggests line printing direction (in case something ends up being able to use that)

a way to be certain about the actual line width when using the font in CAD

##### Ideas for getting there
to do most of this it may help if the tooling could do some of the following:
* save the stroke when doing expansion
* easily delete the expanded stroke without deleting the center line (or just toggle the expansion)
* allow moving an expanded stroke by the stroke line
* set constraints between features
  * shorten stroke if it gets within $global_line_width of itself or another stroke
  * keep 2x$global_line_width between adjacent character features
* switch to an alternative stroke once line width is thin enough to allow for it
  * e.g. switch from double story to single story when the ratio of font size to line width allows it
  * or even switching to a 2 stroke style
* generate a font from a different layer
* export the lines and expanded strokes separately
  
some of that is very similar to behaviour i'm familiar with from KiCad's PCB editor,
where track(s) (segments) have a constant width,
interactive push and shove routing enforces design rules like track to track clearance,
and the equivalent to not pinching off a counter opening to allow a line through is second nature

### Show & Tell
photo from the original project that got me interested in getting this right
![IMG_20211217_161522](https://user-images.githubusercontent.com/30512529/160431390-0a908f51-1204-4d4a-9c9c-501242ec7069.jpg)

i've had a go at doing something rudimentary to get to know fontforge a bit,
unfortunately i only remembered that this was supposed to be small halfway through and continued as-is :)
![filafont_specimen1_cropped_small](https://user-images.githubusercontent.com/30512529/160427139-db432540-c637-42d8-8239-1a922c1c8d49.png)
note the `m` in `jumps` having 2 lines start a bit late due to the start sticking to the nozzle,
some `b`, `d`, `p`, `q` printing in 3 lines instead of 2 due to the slicer not knowing the strokes i started with,
the dot on the `i` in the URL and the top of the `t` in the alphabet missing due to those not staying stuck to the print bed,
and that the surrounding black was done in "concentric" mode to ensure nice filling of counters
(this print was set to 2 wider perimeters)

(note: these are both printed upside down, which is to say, face down on the print bed, i've called this "first layer infill' before)
(both are orange in black PETG, i think the grey is due to the scanner interacting with the fluorencent orange)

here's a test i did with lines and circles in 0.4mm line width (prusa i3, 0.3mm layer height, PLA), the grid is 10mm
![photo_2022-03-28_17-26-50](https://user-images.githubusercontent.com/30512529/160433331-5bb858b9-a63a-4efd-a049-ea5a853334fe.jpg)
![photo_2022-03-28_17-26-52](https://user-images.githubusercontent.com/30512529/160433333-5ff790d1-8384-4efa-bc98-6982b367e448.jpg)
and something similar with 0.8mm
![photo_2022-03-28_17-26-43](https://user-images.githubusercontent.com/30512529/160433329-0b7a59ab-dbcb-430c-9859-3191f0ed0621.jpg)
![photo_2022-03-28_17-26-48](https://user-images.githubusercontent.com/30512529/160433330-a11f14fe-33a2-4a4e-86dc-7c1647e9642b.jpg)
(the blob in the 4th circle is a slicer artifact that should be resolved by now)

**Comment by ctrlcctrlv** [comment]
Wow, cool!

Would this work for any open contour font? https://github.com/ctrlcctrlv/FRBAmericanCursive/blob/0ef2af1d0b68b974e1b0c5180807640876110b29/FRBAmericanCursive.sfd

**Reply by evils** [reply]
i'm not sure what you mean with "work" and "open contour"

FRB cursive bold doesn't fair too well in superslicer, but i don't have the latest version of that (2.3.57.10)
![image](https://user-images.githubusercontent.com/30512529/160573896-436bf3f8-7c75-4431-b627-8e1979167c71.png)
the regular does about as well it seems
![image](https://user-images.githubusercontent.com/30512529/160576497-c775460a-bab7-431a-9b7b-8d48ab63ea27.png)



prusaslicer usually does a bit better, though it currently has a bug elsewhere preventing me from using it (2.4.1)
![image](https://user-images.githubusercontent.com/30512529/160574416-457b584c-6835-4692-bde0-0c2d76b72a1e.png)
size 6 does quite well there, though it's being drawn as 2 lines very close to eachother
![image](https://user-images.githubusercontent.com/30512529/160574798-caee7e90-8b94-41b2-9364-169a23c52b57.png)

regular style (non-bold) basically disappears below size 9
![image](https://user-images.githubusercontent.com/30512529/160575426-6add9562-45aa-4772-8cc2-f8b64bc136a6.png)
and with "detect thin wall", it looks ok, but ends up being quite noisy, and i wouldn't expect it to print well
![image](https://user-images.githubusercontent.com/30512529/160575967-fa69d51f-afe0-400f-b91a-f38379c5141f.png)



i think a cursive font not made with 3D printing in mind will result in more opportunities for the nozzle to hit already printed parts and drag them along, which is usually a critical failure
if one could be made that keeps the constraints of 3D printing in mind (and the stroke order conveyed to the printer),
i suspect it could actually work better than discrete characters, as at the sizes i'm trying to get to, issues around starting and stopping a line become quite significant (as seen in the grey tests)
but for now, discrete lettering is a bit simpler to work with, and probably a bit more readable, which is ultimately my goal in embedding labeling in a 3D print

**Reply by ctrlcctrlv** [reply]
You seem not to have understood me, which is all right. I linked to the version of the font which is not in OpenType format, but rather in _open contour_ format, as would be used by a CNC machine or a pen plotter for example. So, I was imagining you could use it to make your 3D printer follow the curves of the letters in a single line as a pen would do, not attempt to preserve walls or anything.

**Reply by evils** [reply]
ah, ok
that seems to be the same idea as in [Relief-SingleLine](https://github.com/isdat-type/Relief-SingleLine)
the main issue with that is that i currently don't know of any way to convert the line provided by such a font to g-code
i'm fairly sure it's a supported workflow for pen plotters, but i'm only familiar with slicers for 3D printing

(i assumed the "open" in "open contour" was in reference to being open source, rather than describing a non-closed shape)

the overlapping lines issue still stands though, it's not possible to print a line at the same place a line already exists
so even if i got the printer to make the movement described by the font, the print would fail
this is one of the problems i describe above
i'd love to know if there is a solution to this

**Comment by ChiefMikeK** [comment]
IRT not sticking you might try warming up your tray first with a heat lamp or a 100watt incandescent bulb.

**Reply by evils** [reply]
hmm, i probably wouldn't have to resort to a heat lamp
but setting the heated bed to a different temperature for the text section of a print is definitely worth considering, thanks

my concern is more from a font design standpoint
no matter how good the bed adhesion is (and it can be too good), the shorter the line being printed, the more chance of failure
this means dots sometimes don't work, and the start of a line is sometimes missing
(this may be a side-effect of filament retraction between prints, which reduces stringing between characters)
a font specifically aimed at 3D printing should account for this
for example by making dots short lines (an idea i got from Relief-SingleLine),
and making sure a shortened line doesn't make a character ambiguous
(control over print direction would be a good way to minimize that,
and i'm wondering if there's serif/terminals that could be used that help to establish the start of a line)

**Reply by ChiefMikeK** [reply]
my experiences with `g-code` was mostly way back in the 70's and my experience `STL` files is near to none but, I'm thinking that you're looking for is a 
- https://en.wikipedia.org/wiki/Multiple_master_fonts
- that has a good render at small point size.
- On the Internet `archive.org` I've found a few old fonts but never searched for a plotter-font
- also try searching/asking on reddit various subs ie. raspberry pi, maker, fonts, stl etc...

**Reply by evils** [reply]
ah, so that's what the MM thing in FF is
yes, it's slowly becoming clear to me i'm asking if it's possible to draw variable fonts using open source font editors

from what i can tell, this isn't a well supported thing in FontForge, but may be possible

side note: i'm only going through an `STL` phase because 3D printer tooling is very focused on that, and when the font solves enough of the things with 3D printing in mind, the slicers seem to figure out most of the things correctly (line direction being a significant thing i don't think they figure out)
ideally i'd be able to go directly from font to `g-code`
that would however still require telling the slicer where not to print the rest of the 3D model (not an impossible thing, they already allow you to embed another colour model in an existing model)

unfortunately i don't think any existing font solves the core problem

**Comment by evils** [comment]
i've also been pointed to the [Edition Typeface](https://eliashanzer.com/edition/), another single line font
whose `Contextual` form has some serifs and terminals that look about like what i described above
which may help in reducing line start failures

**Reply by ChiefMikeK** [reply]
There's also [G04 G-Code: Pause / Dwell for Precise CNC Timing](https://www.cnccookbook.com/g04-gcode-pause-dwell/ ) you can add just before the start of a problematic line-item
:neckbeard:
